src/logger/mongoose_json_formatter.erl

```diff
@@ -0,0 +1,107 @@
+%%%-------------------------------------------------------------------
+%%% @copyright 2020, Erlang Solutions Ltd.
+%%% @doc
+%%% A module formatting logger reports into JSON objects.
+%%%
+%%% @end
+%%%-------------------------------------------------------------------
+-module(mongoose_json_formatter).
+
+%% API
+-export([format/2]).
+
+-spec format(logger:log_event(), logger:formatter_config()) -> unicode:chardata().
+format(E = #{msg := {report, #{label := {error_logger, _}, format := Format, args := Terms}}}, FConfig) ->
+    format(E#{msg := {report,
+                      #{unstructured_log =>
+                        unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
+           FConfig);
+format(E = #{msg := {string, String}}, FConfig) ->
+    format(E#{msg := {report,
+                      #{unstructured_log =>
+                        unicode:characters_to_binary(io_lib:format(String, []))}}}, FConfig);
+format(E = #{msg := {report, Report}}, FConfig) when is_map(Report) ->
+    NewMap = process_metadata(E),
+    NewReport = maps:merge(Report, NewMap),
+    NewConfig = maps:merge(default_config(), config_correct_depth(FConfig)),
+    Formatted = format_item(NewReport, NewConfig),
+    unicode:characters_to_binary([jiffy:encode(Formatted), "\n"], utf8);
+format(Map = #{msg := {Format, Terms}}, FConfig) ->
+    format(Map#{msg := {report,
+                        #{unstructured_log =>
+                          unicode:characters_to_binary(io_lib:format(Format, Terms))}}},
+           FConfig).
+
+format_item(_Item, _FConfig = #{depth := 0}) ->
+    <<"...">>;
+format_item(Item, FConfig) when is_map(Item) ->
+    ML = [format_item({Key, Val}, FConfig) || {Key, Val} <- maps:to_list(Item)],
+    lists:foldl(fun(Map1, Acc) -> maps:merge(Map1, Acc) end, #{}, ML);
+format_item({K, V}, FConfig = #{depth := Depth}) ->
+    % Keys need to be strings in JSON
+    % We can get a nested structure as a key K here, it needs to be stringified
+    #{all_to_binary(K, FConfig) => format_item(V, FConfig#{depth := Depth - 1})};
+format_item(Item, FConfig = #{depth := Depth}) when is_list(Item) ->
+    case io_lib:printable_unicode_list(Item) of
+        true ->
+            % We want to print strings as strings
+            format_str(Item, FConfig);
+        false ->
+            % And pass lists of objects as lists of objects
+            [format_item(I, FConfig#{depth := Depth - 1}) || I <- Item]
+    end;
+format_item(Item, FConfig) ->
+    all_to_binary(Item, FConfig).
+
+format_time(T) ->
+    TimeString = calendar:system_time_to_rfc3339(T, [{unit, microsecond}]),
+    unicode:characters_to_binary(TimeString).
+
+all_to_binary(S, FConfig) when is_binary(S) ->
+    shorten_binary(S, FConfig);
+all_to_binary(Something, FConfig) ->
+    Chars = format_str(Something, FConfig),
+    unicode:characters_to_binary(Chars, utf8).
+
+shorten_binary(S, #{format_chars_limit := unlimited}) ->
+    S;
+shorten_binary(S, #{format_chars_limit := L}) ->
```

Comment:
after shortening binary, we should check if the new binary is valid unicode.
If it was valid unicode before shortening and started being invalid unicode after shortening, we should increase truncate limit by one and try again.

To check if unicode binary is valid, use:

```erlang
5> unicode:characters_to_binary(<<208,186,208,190,209>>,utf8,utf8).
{incomplete,<<208,186,208,190>>,<<"Ñ">>}
6> unicode:characters_to_binary(<<208,186,208,190,209,130>>,utf8,utf8).
<<208,186,208,190,209,130>>
```

If it does not return binary - unicode is invalid.
And, of course, we should check that binary is valid unicode before passing it to jiffy. Otherwise bad stuff would happen.

Reply:
Woow, didn't think about that, thanks!